cmd/pileup/pileup.go

```diff
@@ -0,0 +1,116 @@
+// Command Group: "SAM Tools"
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"github.com/vertgenlab/gonomics/dna"
+	"github.com/vertgenlab/gonomics/exception"
+	"github.com/vertgenlab/gonomics/fileio"
+	"github.com/vertgenlab/gonomics/sam"
+	"log"
+	"strings"
+)
+
+func usage() {
+	fmt.Print(
+		"pileup - Count bases from sequencing data\n\n" +
+			"pileup prints to stdout by default, but can be redirected with the -o flag.\n" +
```

Comment:
I was never a fan of the decision to avoid stdout, but I think I was outvoted. Generating pileups from bam alignments can be thought of as a form of data wrangling and it's more common (and easier) to use stdout. This meme is mostly a joke, but I think it captures this program design pretty well. Stdout is the quicker user experience in practice, and the well-paved side walk is the fileio (to file solution).

![IMG_1126](https://user-images.githubusercontent.com/11465271/131246016-c1878600-2daa-455e-9869-bf711d8876a0.JPG)

Reply:
I think there are a few things swirling around here and I will do my best to separate them out.

I think there are three general options for writing output:
1) Write it to stdout
2) Write it to stdout, but give an option (e.g. -o) so that it can be redirected within the program to a file
3) Write it to a file, where it can be sent to stdout by giving /dev/stdout (or stdout in our case) as the filename

There are similar options for input, but I will just talk about output, where I think most things I will write are similar for the case of inputs.

If we think about programs with a single input or output file/stream, I don't really think there is a huge difference, but I am open to hearing everyone's thoughts.  The options for the two general cases are of writing to stdout or a file for the three options are:

1)
program in.sam
program in.sam > out.bed
2)
program in.sam
program in.sam -o out.bed
3)
program in.sam /dev/stdout
program in.sam out.bed

I can see how people may have some preferences, but I find them all pretty similar.

While it is not a concern with the command-line interface itself, I do worry that option 1 can tempt us to put print statements deep in functions that print directly to stdout and don't take a handler as an argument.  This can lead to the code being difficult to re-use either as the program evolves over time, or as someone else wants to use it for a different program.

I think we start to see some differentiation in style when we have multiple output files, which while uncommon in typical gnu utilities, does happen at times in genomics software.  In this case I think it is good to use filenames (or at least allow the output to be directed to a file inside the program) so that it is possible to steam any of the inputs or outputs.

While the case of multiple output files seems separate, there have been times when a program first envisioned as a small one-output program, has turned out to be useful for related things and grown to provide additional outputs.  My personal preference for option 3), with option 2) being an equal/close second is not so much about the present, but maintaining the code over longer periods of time.

I think the core of this is how strict we want the style convention to be.  As I stated above, I think most people will have a preference about the three interfaces, but at this point in time I don't think this goes beyond a preference.  Style conventions come into play in areas like this where there are equally-good ways to do things, but the team as a whole can benefit from a uniform approach.  We can discuss at code review if we want a stricter or looser style convention here; I am open to either one and interested to get everyone's input.

Reply:
Thank you for the insightful comments! I won't be in attendance of the discussion, but I'll add my two cents here. There is a way (in golang) around the pitfalls of option 1; one can use `log.Print()` and `fmt.Print()` to differentiate print statements from data results. However, I think I would prefer options 2 and/or 3 in most cases for the reasons @craiglowe mentioned.

What I push back on is being **_too strict_** about style conventions and/or standards. While I agree that teams benefit from a uniform approach, there is still a creative and artistic aspect to programming. People sacrifice a lot of time anre are often very passionate about their work and I feel like everyone should have the freedom to implement their style preferences (given there are no functional pitfalls).